File: java/Problem-823.java
// Problem - 823

// https://leetcode.com/problems/binary-trees-with-factors/

// O(n^2) time complexity and O(n) space complexity solution

class Solution {
    final double mod = 1e9 + 7;
    
    public int numFactoredBinaryTrees(int[] arr) {
        HashMap<Integer, Long> dp = new HashMap<Integer, Long>();
        int ans = 0;
        Arrays.sort(arr);
        for(int i = 0; i < arr.length; i++) {
            long ctr = 1;
            for(int j = 0; j < i; j++) {
                if(arr[i] % arr[j] == 0 && dp.containsKey(arr[i]/arr[j])) {
                    ctr = (ctr + (dp.get(arr[j]) * dp.get(arr[i]/arr[j]))%(int)mod)%(int)mod;
                }
            }
            dp.put(arr[i], ctr);
            ans = (ans + (int)ctr)%(int)mod;
        }
        return ans;
    }
}

File: Problem-823.cpp
// Problem - 823

// https://leetcode.com/problems/binary-trees-with-factors/

// O(n^2) time complexity and O(1) space complexity solution where n = no. of chars

class Solution {
public:
    const int mod = 1e9 + 7;
    
    int numFactoredBinaryTrees(vector<int>& arr) {
        unordered_map <int, long> dp;
        int ans = 0;
        sort(arr.begin(), arr.end());
        for(int i = 0; i < arr.size(); i++) {
            long ctr = 1;
            for(int j = 0; j < i; j++) {
                if(arr[i] % arr[j] == 0) {
                    ctr = (ctr + (dp[arr[j]] * dp[arr[i]/arr[j]])%mod)%mod;
                }
            }
            dp[arr[i]] = ctr;
            ans = (ans + ctr)%mod;
        }
        return ans;
    }
};
